### PR TITLE
[PSM Interop] Add "hook service"

### PR DESCRIPTION
**Pull request by eugeneo**

### CODE REVIEW

**Review by yashykt — COMMENTED** [review]
probably better to use callback api :D

**Comment by eugeneo** [comment]
Done. PTAL.

File: test/cpp/interop/pre_stop_hook_server.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#include <grpcpp/server.h>
+
+#include "src/core/lib/config/core_configuration.h"
+
+namespace grpc {
+namespace testing {
+
+class PreStopHookServer;
+
+class PreStopHookServerManager {
```
Comment (yashykt) [inline test/cpp/interop/pre_stop_hook_server.h:35]: Can you add some comments here as to what this does please?

File: src/proto/grpc/testing/test.proto [hunk 2]
```diff
@@ -89,10 +91,17 @@ service LoadBalancerStatsService {
       returns (LoadBalancerAccumulatedStatsResponse) {}
 }
 
+// Hook service that may be started on request by calling XdsUpdateHealthService
+// with HookRequestCommand::START
```
Comment (yashykt) [inline src/proto/grpc/testing/test.proto:95]: Let's add a comment on the benefit of the service over here

File: test/cpp/interop/pre_stop_hook_server.cc [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+      pending_status_ = status;
+    }
+    for (auto request : pending_requests_) {
+      request->Finish(status);
+    }
+    pending_requests_.clear();
+    request_var_.SignalAll();
+  }
+
+  bool ExpectRequests(size_t expected_requests_count, size_t timeout_s) {
```
Comment (yashykt) [inline test/cpp/interop/pre_stop_hook_server.cc:64]: s/ExpectRequests/TestOnlyExpectRequests

File: test/cpp/interop/pre_stop_hook_server.cc [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    request_var_.SignalAll();
+    return reactor;
+  }
+
+  void SetReturnStatus(const Status& status) {
+    grpc_core::MutexLock lock(&mu_);
+    if (pending_requests_.empty()) {
+      pending_status_ = status;
+    }
+    for (auto request : pending_requests_) {
```
Comment (yashykt) [inline test/cpp/interop/pre_stop_hook_server.cc:57]: as we discussed, 1 request - 1 status

**Review by yashykt — CHANGES_REQUESTED** [review]

File: test/cpp/interop/pre_stop_hook_server.h (hunk 1 above)
Comment (eugeneo) [reply test/cpp/interop/pre_stop_hook_server.h:39]: Did that.

**Comment by gnossen** [comment]
Adding Sanjay as author of the test spec.

**Review by yashykt — APPROVED** [review]
Let's change the `ExpectRequests` to `TestOnlyExpectRequests`

I would've preferred to see some discussion on the spec and the need for this, but given that this entire thing is just for interop, I'm fine-ish with checking this in and testing it.

**Pull request merged**